File: libs/sdk-typescript/src/Git.ts
/*
 * Copyright 2025 Daytona Platforms Inc.
 * SPDX-License-Identifier: Apache-2.0
 */

import { GitApi } from '@daytona/toolbox-api-client'
import type { ListBranchResponse, ListRemotesResponse, GitStatus } from '@daytona/toolbox-api-client'
import { WithInstrumentation } from './utils/otel.decorator'

/**
 * Response from the git commit.
 *
 * @interface
 * @property {string} sha - The SHA of the commit
 */
export interface GitCommitResponse {
  sha: string
}

/**
 * Provides Git operations within a Sandbox.
 *
 * @class
 */
export class Git {
  constructor(private readonly apiClient: GitApi) {}

  /**
   * Stages the specified files for the next commit, similar to
   * running 'git add' on the command line.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @param {string[]} files - List of file paths or directories to stage, relative to the repository root
   * @returns {Promise<void>}
   *
   * @example
   * // Stage a single file
   * await git.add('workspace/repo', ['file.txt']);
   *
   * @example
   * // Stage whole repository
   * await git.add('workspace/repo', ['.']);
   */
  @WithInstrumentation()
  public async add(path: string, files: string[]): Promise<void> {
    await this.apiClient.addFiles({
      path,
      files,
    })
  }

  /**
   * List branches in the repository.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @returns {Promise<ListBranchResponse>} List of branches in the repository
   *
   * @example
   * const response = await git.branches('workspace/repo');
   * console.log(`Branches: ${response.branches}`);
   */
  @WithInstrumentation()
  public async branches(path: string): Promise<ListBranchResponse> {
    const response = await this.apiClient.listBranches(path)
    return response.data
  }

  /**
   * Create branch in the repository.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @param {string} name - Name of the new branch to create
   * @returns {Promise<void>}
   *
   * @example
   * await git.createBranch('workspace/repo', 'new-feature');
   */
  @WithInstrumentation()
  public async createBranch(path: string, name: string): Promise<void> {
    await this.apiClient.createBranch({
      path,
      name,
    })
    return
  }

  /**
   * Delete branche in the repository.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @param {string} name - Name of the branch to delete
   * @returns {Promise<void>}
   *
   * @example
   * await git.deleteBranch('workspace/repo', 'new-feature');
   */
  @WithInstrumentation()
  public async deleteBranch(path: string, name: string): Promise<void> {
    await this.apiClient.deleteBranch({
      path,
      name,
    })
    return
  }

  /**
   * Checkout branche in the repository.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @param {string} branch - Name of the branch to checkout
   * @returns {Promise<void>}
   *
   * @example
   * await git.checkoutBranch('workspace/repo', 'new-feature');
   */
  @WithInstrumentation()
  public async checkoutBranch(path: string, branch: string): Promise<void> {
    await this.apiClient.checkoutBranch({
      path,
      branch,
    })
    return
  }

  /**
   * Clones a Git repository into the specified path. It supports
   * cloning specific branches or commits, and can authenticate with the remote
   * repository if credentials are provided.
   *
   * @param {string} url - Repository URL to clone from
   * @param {string} path - Path where the repository should be cloned. Relative paths are resolved based on the sandbox working directory.
   * @param {string} [branch] - Specific branch to clone. If not specified, clones the default branch
   * @param {string} [commitId] - Specific commit to clone. If specified, the repository will be left in a detached HEAD state at this commit
   * @param {string} [username] - Git username for authentication
   * @param {string} [password] - Git password or token for authentication
   * @param {boolean} [insecureSkipTls] - Skip TLS certificate verification (insecure). Use only for trusted internal Git servers with self-signed or private-CA certs.
   * @param {number} [depth] - Create a shallow clone truncated to the given number of commits.
   * @returns {Promise<void>}
   *
   * @example
   * // Clone the default branch
   * await git.clone(
   *   'https://github.com/user/repo.git',
   *   'workspace/repo'
   * );
   *
   * @example
   * // Clone a specific branch with authentication
   * await git.clone(
   *   'https://github.com/user/private-repo.git',
   *   'workspace/private',
   *   branch='develop',
   *   username='user',
   *   password='token'
   * );
   *
   * @example
   * // Clone a specific commit
   * await git.clone(
   *   'https://github.com/user/repo.git',
   *   'workspace/repo-old',
   *   commitId='abc123'
   * );
   */
  @WithInstrumentation()
  public async clone(
    url: string,
    path: string,
    branch?: string,
    commitId?: string,
    username?: string,
    password?: string,
    insecureSkipTls?: boolean,
    depth?: number,
  ): Promise<void> {
    await this.apiClient.cloneRepository({
      url: url,
      branch: branch,
      path,
      username,
      password,
      commit_id: commitId,
      insecure_skip_tls: insecureSkipTls,
      depth,
    })
  }

  /**
   * Commits staged changes.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @param {string} message - Commit message describing the changes
   * @param {string} author - Name of the commit author
   * @param {string} email - Email address of the commit author
   * @param {boolean} [allowEmpty] - Allow creating an empty commit when no changes are staged
   * @returns {Promise<void>}
   *
   * @example
   * // Stage and commit changes
   * await git.add('workspace/repo', ['README.md']);
   * await git.commit(
   *   'workspace/repo',
   *   'Update documentation',
   *   'John Doe',
   *   'john@example.com',
   *   true
   * );
   *
   */
  @WithInstrumentation()
  public async commit(
    path: string,
    message: string,
    author: string,
    email: string,
    allowEmpty?: boolean,
  ): Promise<GitCommitResponse> {
    const response = await this.apiClient.commitChanges({
      path,
      message,
      author,
      email,
      allow_empty: allowEmpty,
    })
    return {
      sha: response.data.hash,
    }
  }

  /**
   * Push local changes to the remote repository.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @param {string} [username] - Git username for authentication
   * @param {string} [password] - Git password or token for authentication
   * @param {string} [branch] - Branch to push. Defaults to the current branch
   * @param {string} [remote] - Remote to push to. Defaults to "origin"
   * @param {boolean} [setUpstream] - Record the pushed branch as the upstream tracking branch
   * @returns {Promise<void>}
   *
   * @example
   * // Push to a public repository
   * await git.push('workspace/repo');
   *
   * @example
   * // Push to a private repository
   * await git.push(
   *   'workspace/repo',
   *   'user',
   *   'token'
   * );
   *
   * @example
   * // Push a new branch and set its upstream
   * await git.push('workspace/repo', undefined, undefined, 'feature', undefined, true);
   */
  @WithInstrumentation()
  public async push(
    path: string,
    username?: string,
    password?: string,
    branch?: string,
    remote?: string,
    setUpstream?: boolean,
  ): Promise<void> {
    await this.apiClient.pushChanges({
      path,
      username,
      password,
      branch,
      remote,
      set_upstream: setUpstream,
    })
  }

  /**
   * Pulls changes from the remote repository.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @param {string} [username] - Git username for authentication
   * @param {string} [password] - Git password or token for authentication
   * @param {string} [branch] - Branch to pull. Defaults to the current branch's upstream
   * @param {string} [remote] - Remote to pull from. Defaults to "origin"
   * @returns {Promise<void>}
   *
   * @example
   * // Pull from a public repository
   * await git.pull('workspace/repo');
   *
   * @example
   * // Pull from a private repository
   * await git.pull(
   *   'workspace/repo',
   *   'user',
   *   'token'
   * );
   *
   * @example
   * // Pull a specific branch from a specific remote
   * await git.pull('workspace/repo', undefined, undefined, 'main', 'upstream');
   */
  @WithInstrumentation()
  public async pull(
    path: string,
    username?: string,
    password?: string,
    branch?: string,
    remote?: string,
  ): Promise<void> {
    await this.apiClient.pullChanges({
      path,
      username,
      password,
      branch,
      remote,
    })
  }

  /**
   * Gets the current status of the Git repository.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @returns {Promise<GitStatus>} Current repository status including:
   *                               - currentBranch: Name of the current branch
   *                               - ahead: Number of commits ahead of the remote branch
   *                               - behind: Number of commits behind the remote branch
   *                               - branchPublished: Whether the branch has been published to the remote repository
   *                               - fileStatus: List of file statuses
   *
   * @example
   * const status = await sandbox.git.status('workspace/repo');
   * console.log(`Current branch: ${status.currentBranch}`);
   * console.log(`Commits ahead: ${status.ahead}`);
   * console.log(`Commits behind: ${status.behind}`);
   */
  @WithInstrumentation()
  public async status(path: string): Promise<GitStatus> {
    const response = await this.apiClient.getStatus(path)
    return response.data
  }

  /**
   * Initializes a new Git repository at the specified path.
   *
   * @param {string} path - Path where the repository should be initialized. Relative paths are resolved based on the sandbox working directory.
   * @param {boolean} [bare] - Create a bare repository without a working tree
   * @param {string} [initialBranch] - Name of the initial branch. If not specified, uses the Git default
   * @returns {Promise<void>}
   *
   * @example
   * await git.init('workspace/repo', false, 'main');
   */
  @WithInstrumentation()
  public async init(path: string, bare?: boolean, initialBranch?: string): Promise<void> {
    await this.apiClient.initRepository({
      path,
      bare,
      initial_branch: initialBranch,
    })
  }

  /**
   * Resets the current HEAD to the specified state.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @param {string} [mode] - Reset mode, one of "soft", "mixed" (default), "hard", "merge" or "keep"
   * @param {string} [target] - Revision to reset to. Defaults to HEAD
   * @param {string[]} [files] - Constrain the reset to the given paths
   * @returns {Promise<void>}
   *
   * @example
   * // Unstage all changes (mixed reset to HEAD)
   * await git.reset('workspace/repo');
   *
   * @example
   * // Hard reset to a previous commit
   * await git.reset('workspace/repo', 'hard', 'HEAD~1');
   */
  @WithInstrumentation()
  public async reset(path: string, mode?: string, target?: string, files?: string[]): Promise<void> {
    await this.apiClient.resetChanges({
      path,
      mode,
      target,
      files,
    })
  }

  /**
   * Restores working tree files or unstages changes.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @param {string[]} files - File paths to restore
   * @param {boolean} [staged] - Restore the staging index for the given files
   * @param {boolean} [worktree] - Restore the working tree for the given files. Defaults to true when neither staged nor worktree is provided
   * @param {string} [source] - Restore file contents from the given revision instead of the index
   * @returns {Promise<void>}
   *
   * @example
   * // Discard working tree changes
   * await git.restore('workspace/repo', ['file.txt']);
   *
   * @example
   * // Unstage changes
   * await git.restore('workspace/repo', ['file.txt'], true);
   */
  @WithInstrumentation()
  public async restore(
    path: string,
    files: string[],
    staged?: boolean,
    worktree?: boolean,
    source?: string,
  ): Promise<void> {
    await this.apiClient.restoreFiles({
      path,
      files,
      staged,
      worktree,
      source,
    })
  }

  /**
   * Adds (or overwrites) a remote in the repository.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @param {string} name - Name of the remote
   * @param {string} url - URL of the remote
   * @param {boolean} [fetch] - Fetch from the remote immediately after adding it
   * @param {boolean} [overwrite] - Replace an existing remote with the same name
   * @returns {Promise<void>}
   *
   * @example
   * await git.remoteAdd('workspace/repo', 'origin', 'https://github.com/user/repo.git');
   */
  @WithInstrumentation()
  public async remoteAdd(path: string, name: string, url: string, fetch?: boolean, overwrite?: boolean): Promise<void> {
    await this.apiClient.addRemote({
      path,
      name,
      url,
      fetch,
      overwrite,
    })
  }

  /**
   * Lists the remotes configured in the repository.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @returns {Promise<ListRemotesResponse>} The configured remotes (name + URL)
   *
   * @example
   * const response = await git.remotes('workspace/repo');
   * response.remotes.forEach((r) => console.log(`${r.name}: ${r.url}`));
   */
  @WithInstrumentation()
  public async remotes(path: string): Promise<ListRemotesResponse> {
    const response = await this.apiClient.listRemotes(path)
    return response.data
  }

  /**
   * Gets the URL of a remote, or undefined when it does not exist.
   *
   * @param {string} path - Path to the Git repository root. Relative paths are resolved based on the sandbox working directory.
   * @param {string} name - Name of the remote
   * @returns {Promise<string | undefined>} The remote URL, or undefined when the remote does not exist
   *
   * @example
   * const url = await git.remoteGet('workspace/repo', 'origin');
   */
  @WithInstrumentation()
  public async remoteGet(path: string, name: string): Promise<string | undefined> {
    const response = await this.apiClient.listRemotes(path)
    return response.data.remotes.find((r) => r.name === name)?.url
  }

  /**
   * Sets a Git config value at the given scope.
   *
   * @param {string} key - Config key in dotted form (e.g. "user.name")
   * @param {string} value - Config value
   * @param {string} [scope] - Config scope, one of "global" (default), "local" or "system"
   * @param {string} [path] - Repository path, required when scope is "local"
   * @returns {Promise<void>}
   *
   * @example
   * await git.setConfig('user.name', 'John Doe');
   */
  @WithInstrumentation()
  public async setConfig(key: string, value: string, scope = 'global', path?: string): Promise<void> {
    await this.apiClient.setGitConfig({
      key,
      value,
      scope,
      path,
    })
  }

  /**
   * Gets a Git config value at the given scope, or undefined when unset.
   *
   * @param {string} key - Config key in dotted form (e.g. "user.name")
   * @param {string} [scope] - Config scope, one of "global" (default), "local" or "system"
   * @param {string} [path] - Repository path, required when scope is "local"
   * @returns {Promise<string | undefined>} The config value, or undefined when the key is not set
   *
   * @example
   * const name = await git.getConfig('user.name');
   */
  @WithInstrumentation()
  public async getConfig(key: string, scope = 'global', path?: string): Promise<string | undefined> {
    const response = await this.apiClient.getGitConfig(key, path, scope)
    return response.data.value
  }

  /**
   * Configures the Git user name and email at the given scope.
   *
   * @param {string} name - User name (user.name)
   * @param {string} email - User email (user.email)
   * @param {string} [scope] - Config scope, one of "global" (default), "local" or "system"
   * @param {string} [path] - Repository path, required when scope is "local"
   * @returns {Promise<void>}
   *
   * @example
   * await git.configureUser('John Doe', 'john@example.com');
   */
  @WithInstrumentation()
  public async configureUser(name: string, email: string, scope = 'global', path?: string): Promise<void> {
    await this.apiClient.configureUser({
      name,
      email,
      scope,
      path,
    })
  }

  /**
   * Persists Git credentials globally so that subsequent operations against the
   * given host authenticate automatically.
   *
   * @remarks This stores the password in plaintext on disk via the Git credential store.
   *
   * @param {string} username - Git username
   * @param {string} password - Git password or token
   * @param {string} [host] - Host to authenticate against. Defaults to "github.com"
   * @param {string} [protocol] - Protocol to authenticate against. Defaults to "https"
   * @returns {Promise<void>}
   *
   * @example
   * await git.dangerouslyAuthenticate('user', 'github_token');
   */
  @WithInstrumentation()
  public async dangerouslyAuthenticate(
    username: string,
    password: string,
    host?: string,
    protocol?: string,
  ): Promise<void> {
    await this.apiClient.authenticate({
      username,
      password,
      host,
      protocol,
    })
  }
}

File: libs/sdk-go/pkg/options/git.go
// Copyright 2025 Daytona Platforms Inc.
// SPDX-License-Identifier: Apache-2.0

// Package options provides functional option types for configuring SDK operations.
//
// This package uses the functional options pattern to provide a clean, extensible
// API for configuring optional parameters. Each option function returns a closure
// that modifies the corresponding options struct.
//
// # Usage
//
// Options are passed as variadic arguments to SDK methods:
//
//	err := sandbox.Git.Clone(ctx, url, path,
//	    options.WithBranch("develop"),
//	    options.WithUsername("user"),
//	    options.WithPassword("token"),
//	)
//
// # Generic Apply Function
//
// The [Apply] function creates a new options struct and applies all provided
// option functions to it:
//
//	opts := options.Apply(
//	    options.WithBranch("main"),
//	    options.WithUsername("user"),
//	)
//	// opts.Branch == "main", opts.Username == "user"
package options

// Apply creates a new instance of type T and applies all provided option functions.
//
// This generic function enables a consistent pattern for applying functional options
// across different option types. It allocates a zero-value instance of T, then
// applies each option function in order.
//
// Example:
//
//	opts := options.Apply(
//	    options.WithBranch("main"),
//	    options.WithUsername("user"),
//	)
func Apply[T any](opts ...func(*T)) *T {
	result := new(T)
	for _, opt := range opts {
		opt(result)
	}
	return result
}

// GitClone holds optional parameters for [daytona.GitService.Clone].
//
// Fields are pointers to distinguish between unset values and zero values.
// Use the corresponding With* functions to set these options.
type GitClone struct {
	Branch          *string // Branch to clone (defaults to repository's default branch)
	CommitId        *string // Specific commit SHA to checkout after cloning
	Username        *string // Username for HTTPS authentication
	Password        *string // Password or token for HTTPS authentication
	InsecureSkipTLS *bool   // Skip TLS certificate verification (insecure). Use only for trusted internal Git servers with self-signed or private-CA certs.
	Depth           *int32  // Create a shallow clone truncated to the given number of commits
}

// WithDepth creates a shallow clone truncated to the given number of commits.
//
// Example:
//
//	err := sandbox.Git.Clone(ctx, url, path, options.WithDepth(1))
func WithDepth(depth int32) func(*GitClone) {
	return func(opts *GitClone) {
		opts.Depth = &depth
	}
}

// WithBranch sets the branch to clone instead of the repository's default branch.
//
// Example:
//
//	err := sandbox.Git.Clone(ctx, url, path, options.WithBranch("develop"))
func WithBranch(branch string) func(*GitClone) {
	return func(opts *GitClone) {
		opts.Branch = &branch
	}
}

// WithCommitId sets a specific commit SHA to checkout after cloning.
//
// The repository is first cloned, then the specified commit is checked out,
// resulting in a detached HEAD state.
//
// Example:
//
//	err := sandbox.Git.Clone(ctx, url, path, options.WithCommitId("abc123def"))
func WithCommitId(commitID string) func(*GitClone) {
	return func(opts *GitClone) {
		opts.CommitId = &commitID
	}
}

// WithUsername sets the username for HTTPS authentication when cloning.
//
// For GitHub, GitLab, and similar services, the username is typically your
// account username or a placeholder like "git" when using tokens.
//
// Example:
//
//	err := sandbox.Git.Clone(ctx, url, path,
//	    options.WithUsername("username"),
//	    options.WithPassword("github_token"),
//	)
func WithUsername(username string) func(*GitClone) {
	return func(opts *GitClone) {
		opts.Username = &username
	}
}

// WithPassword sets the password or access token for HTTPS authentication when cloning.
//
// For GitHub, use a Personal Access Token (PAT). For GitLab, use a Project
// Access Token or Personal Access Token. For Bitbucket, use an App Password.
//
// Example:
//
//	err := sandbox.Git.Clone(ctx, url, path,
//	    options.WithUsername("username"),
//	    options.WithPassword("ghp_xxxxxxxxxxxx"),
//	)
func WithPassword(password string) func(*GitClone) {
	return func(opts *GitClone) {
		opts.Password = &password
	}
}

// WithInsecureSkipTLS opts into skipping TLS certificate verification for the
// clone. Use ONLY when cloning from a trusted internal Git server with a
// self-signed or private-CA certificate. Credentials, if supplied, will be
// transmitted over an unverified TLS connection and are exposed to any MITM
// on the route.
//
// Example:
//
//	err := sandbox.Git.Clone(ctx, url, path,
//	    options.WithInsecureSkipTLS(true),
//	)
func WithInsecureSkipTLS(insecureSkipTLS bool) func(*GitClone) {
	return func(opts *GitClone) {
		opts.InsecureSkipTLS = &insecureSkipTLS
	}
}

// GitCommit holds optional parameters for [daytona.GitService.Commit].
type GitCommit struct {
	AllowEmpty *bool // Allow creating commits with no staged changes
}

// WithAllowEmpty allows creating a commit even when there are no staged changes.
//
// This is useful for triggering CI/CD pipelines or marking points in history
// without actual code changes.
//
// Example:
//
//	resp, err := sandbox.Git.Commit(ctx, path, "Trigger rebuild", author, email,
//	    options.WithAllowEmpty(true),
//	)
func WithAllowEmpty(allowEmpty bool) func(*GitCommit) {
	return func(opts *GitCommit) {
		opts.AllowEmpty = &allowEmpty
	}
}

// GitDeleteBranch holds optional parameters for [daytona.GitService.DeleteBranch].
type GitDeleteBranch struct {
	Force *bool // Force delete even if branch is not fully merged
}

// WithForce enables force deletion of a branch even if it's not fully merged.
//
// Use with caution as this can result in lost commits if the branch contains
// work that hasn't been merged elsewhere.
//
// Example:
//
//	err := sandbox.Git.DeleteBranch(ctx, path, "feature/abandoned",
//	    options.WithForce(true),
//	)
func WithForce(force bool) func(*GitDeleteBranch) {
	return func(opts *GitDeleteBranch) {
		opts.Force = &force
	}
}

// GitPush holds optional parameters for [daytona.GitService.Push].
type GitPush struct {
	Username    *string // Username for HTTPS authentication
	Password    *string // Password or token for HTTPS authentication
	Branch      *string // Branch to push (defaults to the current branch)
	Remote      *string // Remote to push to (defaults to "origin")
	SetUpstream *bool   // Record the pushed branch as the upstream tracking branch
}

// WithPushBranch sets the branch to push instead of the current branch.
func WithPushBranch(branch string) func(*GitPush) {
	return func(opts *GitPush) {
		opts.Branch = &branch
	}
}

// WithPushRemote sets the remote to push to instead of "origin".
func WithPushRemote(remote string) func(*GitPush) {
	return func(opts *GitPush) {
		opts.Remote = &remote
	}
}

// WithSetUpstream records the pushed branch as the upstream tracking branch
// (git push --set-upstream).
func WithSetUpstream(setUpstream bool) func(*GitPush) {
	return func(opts *GitPush) {
		opts.SetUpstream = &setUpstream
	}
}

// WithPushUsername sets the username for HTTPS authentication when pushing.
//
// Example:
//
//	err := sandbox.Git.Push(ctx, path,
//	    options.WithPushUsername("username"),
//	    options.WithPushPassword("github_token"),
//	)
func WithPushUsername(username string) func(*GitPush) {
	return func(opts *GitPush) {
		opts.Username = &username
	}
}

// WithPushPassword sets the password or access token for HTTPS authentication when pushing.
//
// Example:
//
//	err := sandbox.Git.Push(ctx, path,
//	    options.WithPushUsername("username"),
//	    options.WithPushPassword("ghp_xxxxxxxxxxxx"),
//	)
func WithPushPassword(password string) func(*GitPush) {
	return func(opts *GitPush) {
		opts.Password = &password
	}
}

// GitPull holds optional parameters for [daytona.GitService.Pull].
type GitPull struct {
	Username *string // Username for HTTPS authentication
	Password *string // Password or token for HTTPS authentication
	Branch   *string // Branch to pull (defaults to the current branch's upstream)
	Remote   *string // Remote to pull from (defaults to "origin")
}

// WithPullBranch sets the branch to pull instead of the current branch's upstream.
func WithPullBranch(branch string) func(*GitPull) {
	return func(opts *GitPull) {
		opts.Branch = &branch
	}
}

// WithPullRemote sets the remote to pull from instead of "origin".
func WithPullRemote(remote string) func(*GitPull) {
	return func(opts *GitPull) {
		opts.Remote = &remote
	}
}

// WithPullUsername sets the username for HTTPS authentication when pulling.
//
// Example:
//
//	err := sandbox.Git.Pull(ctx, path,
//	    options.WithPullUsername("username"),
//	    options.WithPullPassword("github_token"),
//	)
func WithPullUsername(username string) func(*GitPull) {
	return func(opts *GitPull) {
		opts.Username = &username
	}
}

// WithPullPassword sets the password or access token for HTTPS authentication when pulling.
//
// Example:
//
//	err := sandbox.Git.Pull(ctx, path,
//	    options.WithPullUsername("username"),
//	    options.WithPullPassword("ghp_xxxxxxxxxxxx"),
//	)
func WithPullPassword(password string) func(*GitPull) {
	return func(opts *GitPull) {
		opts.Password = &password
	}
}

// GitInit holds optional parameters for [daytona.GitService.Init].
type GitInit struct {
	Bare          *bool   // Create a bare repository without a working tree
	InitialBranch *string // Name of the initial branch (defaults to the Git default)
}

// WithBare creates a bare repository without a working tree.
func WithBare(bare bool) func(*GitInit) {
	return func(opts *GitInit) {
		opts.Bare = &bare
	}
}

// WithInitialBranch sets the name of the initial branch.
func WithInitialBranch(initialBranch string) func(*GitInit) {
	return func(opts *GitInit) {
		opts.InitialBranch = &initialBranch
	}
}

// GitReset holds optional parameters for [daytona.GitService.Reset].
type GitReset struct {
	Mode   *string  // Reset mode: "soft", "mixed" (default), "hard", "merge" or "keep"
	Target *string  // Revision to reset to (defaults to HEAD)
	Files  []string // Constrain the reset to the given paths
}

// WithResetMode sets the reset mode ("soft", "mixed", "hard", "merge" or "keep").
func WithResetMode(mode string) func(*GitReset) {
	return func(opts *GitReset) {
		opts.Mode = &mode
	}
}

// WithResetTarget sets the revision to reset to.
func WithResetTarget(target string) func(*GitReset) {
	return func(opts *GitReset) {
		opts.Target = &target
	}
}

// WithResetFiles constrains the reset to the given paths.
func WithResetFiles(files []string) func(*GitReset) {
	return func(opts *GitReset) {
		opts.Files = files
	}
}

// GitRestore holds optional parameters for [daytona.GitService.Restore].
type GitRestore struct {
	Staged   *bool   // Restore the staging index for the given files
	Worktree *bool   // Restore the working tree for the given files (default when neither is set)
	Source   *string // Restore file contents from the given revision instead of the index
}

// WithRestoreStaged restores the staging index for the given files.
func WithRestoreStaged(staged bool) func(*GitRestore) {
	return func(opts *GitRestore) {
		opts.Staged = &staged
	}
}

// WithRestoreWorktree restores the working tree for the given files.
func WithRestoreWorktree(worktree bool) func(*GitRestore) {
	return func(opts *GitRestore) {
		opts.Worktree = &worktree
	}
}

// WithRestoreSource restores file contents from the given revision instead of the index.
func WithRestoreSource(source string) func(*GitRestore) {
	return func(opts *GitRestore) {
		opts.Source = &source
	}
}

// GitRemoteAdd holds optional parameters for [daytona.GitService.RemoteAdd].
type GitRemoteAdd struct {
	Fetch     *bool // Fetch from the remote immediately after adding it
	Overwrite *bool // Replace an existing remote with the same name
}

// WithRemoteFetch fetches from the remote immediately after adding it.
func WithRemoteFetch(fetch bool) func(*GitRemoteAdd) {
	return func(opts *GitRemoteAdd) {
		opts.Fetch = &fetch
	}
}

// WithRemoteOverwrite replaces an existing remote with the same name.
func WithRemoteOverwrite(overwrite bool) func(*GitRemoteAdd) {
	return func(opts *GitRemoteAdd) {
		opts.Overwrite = &overwrite
	}
}

// GitConfig holds optional parameters for the git config operations
// ([daytona.GitService.SetConfig], [daytona.GitService.GetConfig] and
// [daytona.GitService.ConfigureUser]).
type GitConfig struct {
	Scope *string // Config scope: "global" (default), "local" or "system"
	Path  *string // Repository path, required when scope is "local"
}

// WithConfigScope sets the config scope ("global", "local" or "system").
func WithConfigScope(scope string) func(*GitConfig) {
	return func(opts *GitConfig) {
		opts.Scope = &scope
	}
}

// WithConfigPath sets the repository path (required for the "local" scope).
func WithConfigPath(path string) func(*GitConfig) {
	return func(opts *GitConfig) {
		opts.Path = &path
	}
}

// GitAuthenticate holds optional parameters for [daytona.GitService.DangerouslyAuthenticate].
type GitAuthenticate struct {
	Host     *string // Host to authenticate against (defaults to "github.com")
	Protocol *string // Protocol to authenticate against (defaults to "https")
}

// WithAuthHost sets the host to authenticate against.
func WithAuthHost(host string) func(*GitAuthenticate) {
	return func(opts *GitAuthenticate) {
		opts.Host = &host
	}
}

// WithAuthProtocol sets the protocol to authenticate against.
func WithAuthProtocol(protocol string) func(*GitAuthenticate) {
	return func(opts *GitAuthenticate) {
		opts.Protocol = &protocol
	}
}
